WordPress/gutenberg
# Blocks: Support reusable nested blocks (reusable blocks refactor)

Supersedes #5010
Cherry-picks 840fbec793462039b3cd3279ae807516a0a1a27c and ee61e18cc9c846f032ba663c615b5f9716eac128 from #5010 
Cherry-picks dde17871bd7a57956c0ada10f1df7a8fd7dda117 and 80e03deba7d7737099eed1cdd02e6cc352792c25 from #5223
Cherry-picks 66fdda52ec7177c1a8154e2df2a75da852c813ca from #5220

This pull request seeks to support reusable nested blocks. Unlike #5010, the changes here do not expand the reusable blocks API to support nested blocks, but rather refactor reusable blocks to serve as pointers to the original block references, therefore having better natural support for nested blocks.

A few of the changes have been brought along from #5010:

1. Move context assignment of inner block list from `BlockListBlock` to `BlockEdit`, as the latter is the component used in rendering from `core/block`'s `edit`
   - Downside is a circular dependency from `blocks` to `editor`, as `BlockEdit` requires access to the `BlockList` component
2. Update reusable block UI behaviors to not abort editing when the block becomes deselected, as this is incompatible with nested blocks where selection changes to the inner block

A complication of reusable blocks as pointers it that we need to add support for the blocks state to receive arbitrary blocks after initial load without dirtying state (the loaded result of a reusable block should be merged into but not dirtying to blocks state). New options of `ignoreTypes` have been added to both the `withChangeDetection` and `withHistory` higher-order reducers to accommodate this.

__Testing instructions:__

Verify there are no regressions in the behavior of reusable blocks.

Ensure that you can save a Columns block, including content, as a reusable block.

Ensure unit tests pass:

```
npm test
```

__Remaining tasks:__

- [x] Reusable blocks effects tests
